### ISSUE TITLE
Poller crashes (status "execution expired")

### ISSUE THREAD

**Issue by Nemental** [blindsidenetworks/scalelite#647]
Hi,

yesterday I updated Scalelite to the latest version. Since then the poller crashes with the following logs at random point.

**Poller container logs**
```
W, [2021-08-27T09:57:16.988054 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
W, [2021-08-27T09:57:23.797328 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
W, [2021-08-27T09:57:46.994622 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:57:46.995198 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:57:46.994819 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:57:49.093409 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
W, [2021-08-27T09:57:54.899023 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
W, [2021-08-27T09:58:19.027283 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:58:19.027857 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:58:19.127374 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:58:19.327668 #8]  WARN -- : Server id=***** is unhealthy. Panicking and setting offline...
W, [2021-08-27T09:58:19.227383 #8]  WARN -- : Server id=***** is unhealthy. Panicking and setting offline...
W, [2021-08-27T09:58:19.527669 #8]  WARN -- : Server id=***** is unhealthy. Panicking and setting offline...
W, [2021-08-27T09:58:20.929129 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
W, [2021-08-27T09:58:21.130377 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
Clearing Meeting id=*****
W, [2021-08-27T09:58:26.635399 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
WARNING: Could not end meeting id=*****: execution expired
Clearing Meeting id=*****
W, [2021-08-27T09:58:50.864793 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:58:50.865366 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:58:50.865582 #8]  WARN -- : Failed to get server id=***** status: execution expired
W, [2021-08-27T09:58:50.874441 #8]  WARN -- : Server id=***** is unhealthy. Panicking and setting offline...
W, [2021-08-27T09:58:50.874691 #8]  WARN -- : Server id=***** is unhealthy. Panicking and setting offline...
W, [2021-08-27T09:58:50.874785 #8]  WARN -- : Server id=***** is unhealthy. Panicking and setting offline...
W, [2021-08-27T09:58:50.163158 #8]  WARN -- : Error 3 errors
Servers list concurrently modified (ApplicationRedisRecord::ConcurrentModificationError)
/srv/scalelite/app/models/server.rb:105:in `block (2 levels) in save!'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:2415:in `block in watch'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:69:in `block in synchronize'
/usr/lib/ruby/2.7.0/monitor.rb:202:in `synchronize'
/usr/lib/ruby/2.7.0/monitor.rb:202:in `mon_synchronize'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:69:in `synchronize'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:2410:in `watch'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/redis-namespace-1.8.1/lib/redis/namespace.rb:476:in `call_with_namespace'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/redis-namespace-1.8.1/lib/redis/namespace.rb:352:in `block (2 levels) in <class:Namespace>'
/srv/scalelite/app/models/server.rb:85:in `block in save!'
/srv/scalelite/app/models/application_redis_record.rb:144:in `block in with_connection'
/srv/scalelite/lib/redis_store.rb:35:in `block in with_connection'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:63:in `block (2 levels) in with'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:62:in `handle_interrupt'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:62:in `block in with'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:59:in `handle_interrupt'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:59:in `with'
/srv/scalelite/lib/redis_store.rb:35:in `with_connection'
/srv/scalelite/app/models/application_redis_record.rb:143:in `with_connection'
/srv/scalelite/app/models/application_redis_record.rb:147:in `with_connection'
/srv/scalelite/app/models/server.rb:56:in `save!'
/srv/scalelite/lib/tasks/poll.rake:107:in `block (4 levels) in <main>'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/promises.rb:1582:in `evaluate_to'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/promises.rb:1765:in `block in on_resolvable'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:363:in `run_task'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:352:in `block (3 levels) in create_worker'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `loop'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `block (2 levels) in create_worker'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `catch'
/srv/scalelite/vendor/bundle/ruby/2.7.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `block in create_worker'
W, [2021-08-27T09:58:52.164158 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
W, [2021-08-27T09:58:52.164158 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
W, [2021-08-27T09:58:52.164158 #8]  WARN -- : Failed to check meeting id=***** status: execution expired
```

**Scalelite `rake servers` output**
```
id: *****
        url: https://*****/bigbluebutton/api
        secret: *****
        enabled
        load: unavailable
        load multiplier: 1.0
        offline
id: *****
        url: https://*****/bigbluebutton/api
        secret: *****
        cordoned
        load: unavailable
        load multiplier: 1.0
        offline
id: *****
        url: https://*****/bigbluebutton/api
        secret: *****
        cordoned
        load: unavailable
        load multiplier: 1.0
        offline
```

All servers are shown as offline. I can run `rake poll:all` manually, so they are online, but will be shown as offline again after next poll interval.

**Comment by Nemental** [comment]
@pielonet 
```
WEB_CONCURRENCY=not set (default - 0)
POLLER_THREADS=not set (default - 5)
POLL_INTERVAL=15
POLLER_WAIT_TIMEOUT=not set (default - 10)
CONNECT_TIMEOUT=not set (default - 5)
RESPONSE_TIMEOUT=not set (default - 10)
SERVER_UNHEALTHY_THRESHOLD=not set (default - 2)
```

**Comment by pielonet** [comment]
Try to set the following :
```
POLL_INTERVAL=60
WEB_CONCURRENCY=<number-of-cpus-in-your-server>
POLLER_THREADS=10
```

And make sure your BBB servers are reachable from your Scalelite server :
Check BBB API manually : 
```
curl https://<bbb-host-domain>/bigbluebutton/api/
```
You should get a response with SUCCESS status.

**Comment by Nemental** [comment]
@pielonet 
Thank you, I'll try these settings, except `POLL_INTERVAL`. I can't increase the interval to 60, because sometimes we have several dozen backends registered.

**Comment by pielonet** [comment]
The more BBB backends you have, the more you should increase `POLL_INTERVAL` : it is the delay in seconds between two polling tasks.
For example, I have 150 BBB backends registered in Scalelite.
I set `CONNECT_TIMEOUT` to 2 seconds and `POLLER_THREADS` to 10. This way at least 10 backends get polled every 2 seconds. The complete poll task takes roughly  : (150/10)*2=30 seconds to complete. So setting `POLL_INTERVAL` to 60 seconds is secure.

Make your own computation with your personnal figures (i.e `CONNECT_TIMEOUT=5`).

I guess you should increase `POLL_INTERVAL` to 60 seconds for this reason.

To make it obvious executing `time ./bin/rake poll:all` should give you an idea of the time the polling process take to complete. `POLL_INTERVAL` should be significantly greater than this time in seconds.

**Comment by Nemental** [comment]
While testing, I discovered something that confuses me. After removing a registered backend (ID: `188a448c-785a-4d0d-831a-5c01993bbb32`) there is still a key in the redis that points to the same ID.
```
192.168.***.***:****> keys *
1) "scalelite:188a448c-785a-4d0d-831a-5c01993bbb32"
```
If I'm right, this key shouldn't exist after the removal.

We didn't have any problems on previous releases with `POLL_INTERVAL=15` in the past (for months). I think that's not the problem. The poller also crashes with 1-3 servers (tested).

**Comment by Nemental** [comment]
@pielonet Unfortunately, the problem still exists. During the past week I increased `POLL_INTERVAL` to `20` while running 2-3 backends. I can remember, that the poller also crashed before the Scalelite update, but restarted on its own (increased kubernetes restart-count), so i didn't really recognized it in production.

**Comment by defnull** [comment]
Here is my guess: When polling multiple servers in multiple threads, it is likely that `Server.save!` is called concurrently for different servers. The problem is that `Server.save!` may update shared data structures in addition to the `server:#{id}` key, namely `servers`, `server_enabled`, `server_load` or `cordoned_server_load`. If that happens more than once at the same time, the `redis.multi` transaction will fail and produce the error you are seeing in your logs.

**Comment by amalrails** [comment]
@Nemental the `status "execution expired"` error occurs when the bbb-server doesn't respond within the specified CONNECT_TIMEOUT value. @pielonet  had a very similar issue and it was fixed by updating the config values which he has mentioned above. He also add a case where too many concurrent requests where getting rejected by his DNS server, so we had to reduce the `POLLER_THREADS` value. I am assuming this might be something similar to that issue.
